### ISSUE TITLE
ValueError: '/32' does not appear to be an IPv4 or IPv6 network

### ISSUE THREAD

**Issue by bubublik** [jsocol/django-ratelimit#209]
Migrate from version 2.0
In version 3.0.1 got this error on production server:
```
ValueError: '/32' does not appear to be an IPv4 or IPv6 network
ERROR    2020-09-02 17:30:30 django.request	Internal Server Error: /live/
Traceback (most recent call last):
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/ratelimit/decorators.py", line 18, in _wrapped
    ratelimited = is_ratelimited(request=request, group=group, fn=fn,
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/ratelimit/core.py", line 119, in is_ratelimited
    usage = get_usage(request, group, fn, key, rate, method, increment)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/ratelimit/core.py", line 175, in get_usage
    value = _SIMPLE_KEYS[key](request)
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/ratelimit/core.py", line 49, in <lambda>
    'ip': lambda r: ip_mask(r.META['REMOTE_ADDR']),
  File "/home/zaq/.local/share/virtualenvs/wh-gJZcLoMf/lib/python3.8/site-packages/ratelimit/core.py", line 37, in ip_mask
    network = ipaddress.ip_network('{}/{}'.format(ip, mask), strict=False)
  File "/usr/lib/python3.8/ipaddress.py", line 83, in ip_network
    raise ValueError('%r does not appear to be an IPv4 or IPv6 network' %
ValueError: '/32' does not appear to be an IPv4 or IPv6 network
```

Usage:
```
@method_decorator(ratelimit(key='ip', rate='6/m', method='GET', block=True), name='get')
class SearchPastList(ListView):
```

**Comment by jsocol** [comment]
I think that would happen if `REMOTE_ADDR` were empty (c.f. ratelimit/core.py:37). I'm not sure what could cause that. Is there any custom middleware that could change or remove that value?

**Comment by bubublik** [comment]
> I think that would happen if `REMOTE_ADDR` were empty (c.f. ratelimit/core.py:37). I'm not sure what could cause that. Is there any custom middleware that could change or remove that value?
There is no custom middleware. And all works great at localhost and on development server

**Comment by saadmk11** [comment]
Having the same issue, getting the exact same error. It seems like ``REMOTE_ADDR`` is not being set by ``nginx``. Need to investigate more.

Handling this exception with a Warning Maybe a good start otherwise the server crashes.

**Comment by jsocol** [comment]
Thanks, having another example is helpful! Some things to check:
- how are you running the Django process? (Eg gunicorn, uwsgi, etc)
- how is it communicating to nginx? (Eg tcp sockets, Unix sockets, fcgi, etc)
- what are the relevant config values for both of those?

The error thrown here could be a little more obvious, though the existing error made it fairly clear what was happening. I don’t agree that it should just warn, though. You asked it to ratelimit based on IP and it wasn’t possible to determine an IP for this request. Warning would effectively allow this request to avoid ratelimiting.

**Comment by E-RROR** [comment]
Same issue when deploy Django to production and run it with gunicorn sockets and proxy it with nginx, i don't have problem with development mode

**Comment by jsocol** [comment]
Hi @E-RROR, thanks for adding that! When you say “gunicorn sockets” do you mean a unix socket? Any chance you could share the relevant part of your nginx config, the `upstream {}` block or the `proxy_pass` / `uwsgi_pass` / `fastcgi_pass` / etc directive? (@saadmk11 or @bubublik this could be helpful from either of you, too)

**Comment by E-RROR** [comment]
Sure @jsocol 
```
server {
    listen 80;
    server_name server_domain_or_IP;

    location = /favicon.ico { access_log off; log_not_found off; }
    location /static/ {
        root /home/sammy/myprojectdir;
    }

    location / {
        include proxy_params;
        proxy_pass http://unix:/run/gunicorn.sock;
    }
}
```
and yeah i mean unix sockets
also the django app and ratelimit works fine with all the gunicorn and nginx configuration above but when i set the ```DEBUG = False``` the errors occures

**Comment by BarnabasSzabolcs** [comment]
> Sure @jsocol
> 
> ```
> server {
>     listen 80;
>     server_name server_domain_or_IP;
> 
>     location = /favicon.ico { access_log off; log_not_found off; }
>     location /static/ {
>         root /home/sammy/myprojectdir;
>     }
> 
>     location / {
>         include proxy_params;
>         proxy_pass http://unix:/run/gunicorn.sock;
>     }
> }
> ```
> 
> and yeah i mean unix sockets
> also the django app and ratelimit works fine with all the gunicorn and nginx configuration above but when i set the `DEBUG = False` the errors occures

@E-RROR what is the contents of your proxy_params file? (`/etc/nginx/proxy_params` )

**Comment by BarnabasSzabolcs** [comment]
Ok, now I've fixed the problem. I think it would be nice to have ratelimit fix the missing REMOTE_ADDR problem.
Until then, here's the middleware that one needs to add if running behind `gunicorn`:

```
class XForwardedForMiddleware(MiddlewareMixin):
    """
    Sets REMOTE_ADDR correctly to the client's ip, if django is run on a domain socket server.
    ref: https://stackoverflow.com/a/34254843/1031191
    """
    def process_request(self, request):
        if "HTTP_X_FORWARDED_FOR" in request.META:
            request.META["HTTP_X_PROXY_REMOTE_ADDR"] = request.META["REMOTE_ADDR"]
            parts = request.META["HTTP_X_FORWARDED_FOR"].split(",", 1)
            request.META["REMOTE_ADDR"] = parts[0]
```

**Comment by BarnabasSzabolcs** [comment]
Issue is fixed in my pull request:

[](https://github.com/jsocol/django-ratelimit/pull/217)

(The fix is progressive: http_x_forwarded_for is used as client IP if the client IP is not set in request.meta['REMOTE_ADDR'] which is the culprit of this issue).

**Comment by BarnabasSzabolcs** [comment]
@jsocol Ok, now I read in Your documentation:

> Django-Ratelimit is not the correct place to handle reverse proxies and adjust the IP address, and patches dealing with it will not be accepted. There is too much variation in the wild to handle it safely.

So, I think django-ratelimit should throw an exception if REMOTE_ADDR is not present... in debug mode, and in `DEBUG=False`, it should just tag along. (how about that?)

**Comment by jsocol** [comment]
> For me it is also true that the error comes when `DEBUG = True`, but only on my linux server and not in my development environment that does not use gunicorn but `runserver` but it is also proxy_pass-ed via nginx.

That's interesting. I wonder what `runserver` is doing differently there.

> So, I think django-ratelimit should throw an exception if `REMOTE_ADDR` is not present...

I agree. I mean it already does, but it should be a better, more actionable exception.

> in debug mode, and in `DEBUG=False, it should just tag along. (how about that?)

For things like local/dev/debug deployments, [`RATELIMIT_ENABLE`](https://django-ratelimit.readthedocs.io/en/stable/settings.html#ratelimit-enable) is probably the best way to globally disable it. Breaking the behavior of _only_ IP ratelimiting based on other settings feels unexpected.

> This issue is fixed in pull request #217

Thank you for the PR. Per the docs, I'm not going to merge it (as an example of why I maintain this stance, there is a potential vulnerability in it that I'll clear up over there). It might be worth adding an example piece of middleware to the Cookbook section in the docs, though, with appropriate caveats.
